### PR TITLE
feat: 모니터링 활성화 및 P95/P99 리포트, 즉시 알림 추가

### LINKED ISSUE #107: 모니터링 활성화 + P95/P99 리포트 + 즉시 알림 구축
## 배경

AI 모니터링 파이프라인 코드(log-watcher.sh, diagnostic-agent.py, discord-alert.sh 등)가 이미 구현되어 있지만, 로그 파일 출력이 없고 Cron이 설치되지 않아 실제로 동작하지 않는 상태.
모니터링을 활성화하고, P95/P99 응답시간 리포트를 추가하여 스케일 업 판단의 데이터 기반을 마련한다.

---

## 선행 작업 (Rust 코드)

| 파일 | 내용 |
|------|------|
| `Cargo.toml` | `tracing-appender` 추가 |
| `logging.rs` | 로그 파일 출력 추가 (`logs/server.YYYY-MM-DD.log`) |
| `middleware.rs` | `duration_ms` 측정 + 로그 기록 |

→ 이거 하면 기존 log-watcher.sh 파이프라인 자동 동작 (에러 감지 + AI 진단 + Discord + GitHub Issue)

---

## 1단계: 일일 리포트

**신규: `scripts/daily-report.sh`**
- 매일 9시 Cron
- 로그에서 `duration_ms` 추출 → P50/P95/P99 계산
- 총 요청 수, 에러율, CPU, 메모리 집계
- `discord-alert.sh`로 Discord 전송
- AI 안 씀

---

## 2단계: 즉시 알림

**신규: `scripts/perf-alert.sh`**
- 매분 Cron
- CPU > 80% (3분 연속), 메모리 > 85%, ERROR 급증 체크
- 임계값 초과 시 Discord 즉시 알림
- 중복 방지 10분 간격
- AI 안 씀

---

## 3단계: 수동 스케일 업

- 리포트 데이터 보고 사람이 판단
- t3.micro → t3.medium → t3.large
- P95 지속 상승, CPU 평균 50% 이상 유지 시 스케일 업

---

## Cron 최종

```
* * * * *  perf-alert.sh       # 매분 - 즉시 알림
0 9 * * *  daily-report.sh     # 매일 9시 - 일일 리포트
0 9 * * *  log-watcher.sh      # 매일 9시 - 에러 AI 진단
```

---

## 파일 변경 요약

| 파일 | 작업 | 단계 |
|------|------|------|
| `Cargo.toml` | tracing-appender 추가 | 선행 |
| `logging.rs` | 파일 출력 레이어 추가 | 선행 |
| `middleware.rs` | duration_ms 측정 + 로그 | 선행 |
| `daily-report.sh` | 신규 작성 | 1단계 |
| `perf-alert.sh` | 신규 작성 | 2단계 |
| `setup-cron.sh` | 스케줄 추가 | 1~2단계 |

---

## 향후 계획 (이번 범위 아님)

- 4단계: ALB + ASG 자동 스케일 아웃 + 무중단 배포 (트래픽 증가 시 도입)

### PR DESCRIPTION
**Pull request by catturtle123**
## Summary
closes #107

- 로그 파일 출력 추가로 기존 AI 모니터링 파이프라인(log-watcher.sh) 활성화
- 미들웨어에 응답시간(duration_ms) 측정 추가
- 매일 9시 P50/P95/P99 일일 성능 리포트 스크립트 추가 (daily-report.sh)
- 매분 CPU/메모리/에러 즉시 알림 스크립트 추가 (perf-alert.sh)
- Cron 스케줄 통합 관리 (setup-cron.sh)

## 변경 파일

| 파일 | 내용 |
|------|------|
| `Cargo.toml` | `tracing-appender` 의존성 추가 |
| `logging.rs` | 일별 로그 파일 출력 레이어 추가 |
| `middleware.rs` | duration_ms 측정 + 로그 기록 |
| `daily-report.sh` | P50/P95/P99, 요청수, 에러율, CPU/메모리 일일 리포트 |
| `perf-alert.sh` | CPU/메모리/에러 임계값 즉시 알림 (중복 방지 포함) |
| `setup-cron.sh` | 3개 스크립트 Cron 스케줄 등록 |

## Cron 스케줄

```
* * * * *  perf-alert.sh       # 매분 - 즉시 알림
0 9 * * *  daily-report.sh     # 매일 9시 - 일일 리포트
0 9 * * *  log-watcher.sh      # 매일 9시 - 에러 AI 진단
```

## 배포 후 EC2에서 할 일

```bash
./scripts/setup-cron.sh
```

## Test plan

- [x] `cargo build` 통과
- [x] `cargo test` 통과 (392 tests)
- [x] `cargo clippy -- -D warnings` 통과
- [ ] EC2 배포 후 `setup-cron.sh` 실행하여 Cron 등록
- [ ] 로그 파일 생성 확인 (`logs/server.log.YYYY-MM-DD`)
- [ ] Discord 일일 리포트 수신 확인

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added request metrics logging (method, path, duration_ms, status) with level-based completion events for better request visibility.
  * Preserves and returns a log guard to ensure buffered logs are flushed on shutdown; init no longer fails the process on setup.
  * Introduced dual log output: console plus daily rotating JSON log files (configurable log directory).
  * Maintains x-request-id header behavior while improving per-status logging granularity.

### CODE REVIEW

File: codes/server/src/utils/logging.rs [hunk 1]
```diff
@@ -1,27 +1,44 @@
...
 /// 기본값은 `info,server=debug`입니다.
+///
+/// 로그는 stdout과 `logs/` 디렉토리의 일별 파일에 동시 출력됩니다.
+/// 파일명 형식: `server.YYYY-MM-DD.log`
 pub fn init_logging() {
-    // flatten_event(false)로 설정하여 fields 중첩 구조 유지
-    let fmt_layer = fmt::layer()
+    let log_dir = std::env::var("LOG_DIR").unwrap_or_else(|_| "logs".to_string());
+
+    let file_appender = rolling::daily(&log_dir, "server.log");
```
[the remark this answers is not in the document]
Comment (catturtle123) [reply codes/server/src/utils/logging.rs:21]: 수정했는데 해결되었는지 다시 확인해줘

File: codes/server/src/utils/logging.rs [hunk 2]
```diff
@@ -1,27 +1,44 @@
...
+
+    let stdout_layer = fmt::layer()
         .json()
         .with_timer(fmt::time::UtcTime::rfc_3339())
         .with_current_span(true)
```
Comment (jsoonworld) [inline codes/server/src/utils/logging.rs:27]: `tracing-appender`의 `rolling::daily`는 `WorkerGuard`를 반환하는데, 이를 유지하지 않으면 프로세스 종료 시 버퍼링된 로그가 손실될 수 있습니다.

권장 수정:
```rust
pub fn init_logging() -> tracing_appender::non_blocking::WorkerGuard {
    let file_appender = rolling::daily(&log_dir, "server.log");
    let (non_blocking, guard) = tracing_appender::non_blocking(file_appender);
    // ... non_blocking을 writer로 사용
    guard  // main에서 유지해야 함
}
```

또는 현재 구현을 유지하려면, 프로세스가 graceful shutdown 시간이 충분하다면 괜찮을 수도 있습니다.

File: scripts/daily-report.sh [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+CRITICAL_COUNT=$(jq -r 'select(.level == "ERROR") | select(.fields.error_code != null) | .fields.error_code' "$LOG_FILE" 2>/dev/null | grep -cE "(COMMON500|AI5[0-9]{3})" || true)
+
+# CPU 사용률 (현재 스냅샷)
+if [[ "$(uname)" == "Darwin" ]]; then
+    CPU_USAGE=$(top -l 1 | grep "CPU usage" | awk '{print $3}' | tr -d '%')
+else
+    CPU_USAGE=$(top -bn1 | grep "Cpu(s)" | awk '{printf "%.1f", $2 + $4}')
+fi
+
+# 메모리 사용률
```
Comment (jsoonworld) [inline scripts/daily-report.sh:79]: CPU/Memory 수치는 "현재 스냅샷"인데, 이 리포트는 어제($YESTERDAY) 로그를 분석하는 일일 리포트입니다.

시스템 섹션의 CPU/Memory가 어제 데이터가 아닌 현재 값이라는 것이 혼란스러울 수 있습니다. 두 가지 방안 중 선택을 권장합니다:

1. 리포트에서 시스템 섹션 제거
2. 섹션 제목을 "현재 시스템 상태"로 명확히 표기

File: scripts/perf-alert.sh [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    local file="$1"
+    echo "0" > "$file"
+}
+
+# CPU 사용률 체크
+if [[ "$(uname)" == "Darwin" ]]; then
+    CPU_USAGE=$(top -l 1 -n 0 | grep "CPU usage" | awk '{gsub(/%/, "", $3); print $3}')
+else
+    CPU_USAGE=$(top -bn1 | grep "Cpu(s)" | awk '{printf "%.0f", $2 + $4}')
+fi
```
Comment (jsoonworld) [inline scripts/perf-alert.sh:90]: `bc` 명령어가 설치되어 있지 않은 시스템에서는 스크립트가 실패합니다.

bash 정수 비교로 대체 가능합니다:
```bash
CPU_INT=${CPU_USAGE%.*}  # 소수점 제거
if [ "$CPU_INT" -ge "$CPU_THRESHOLD" ]; then
```

또는 스크립트 상단에 bc 설치 확인 로직 추가를 권장합니다.

File: scripts/perf-alert.sh [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+# 최근 1분 ERROR 급증 체크
+TODAY=$(date +%Y-%m-%d)
+LOG_FILE="$LOG_DIR/server.log.$TODAY"
+
+if [ -f "$LOG_FILE" ]; then
+    if [[ "$(uname)" == "Darwin" ]]; then
+        ONE_MIN_AGO=$(date -v-1M -u +%Y-%m-%dT%H:%M)
+    else
+        ONE_MIN_AGO=$(date -u -d "1 minute ago" +%Y-%m-%dT%H:%M)
+    fi
```
Comment (jsoonworld) [inline scripts/perf-alert.sh:138]: 타임스탬프 매칭이 분(minute) 단위까지만 비교하므로, 실행 시점에 따라 최근 1분이 아닌 0~2분 전 에러를 카운트할 수 있습니다.

더 정확한 방법:
```bash
# jq로 timestamp 파싱 후 시간 비교
NOW=$(date +%s)
RECENT_ERRORS=$(jq -r --argjson now "$NOW" \
  'select(.level=="ERROR") | select((.timestamp | fromdateiso8601) > ($now - 60))' \
  "$LOG_FILE" 2>/dev/null | wc -l)
```

현재 구현도 대략적인 모니터링 용도로는 충분합니다.

File: scripts/run-tests.sh [hunk 6]
```diff
@@ -13,9 +13,9 @@ cd "$PROJECT_ROOT/codes/server"
 cargo test 2>&1 | tee /tmp/test-results.log
 
 # 결과 파싱
-PASSED=$(grep -c "ok$" /tmp/test-results.log || echo "0")
-FAILED=$(grep -c "FAILED$" /tmp/test-results.log || echo "0")
-IGNORED=$(grep -c "ignored$" /tmp/test-results.log || echo "0")
+PASSED=$(grep -c "ok$" /tmp/test-results.log || true)
```
Comment (jsoonworld) [inline scripts/run-tests.sh:16]: 좋은 수정입니다! `grep -c`는 매칭이 없으면 exit code 1을 반환하므로 `set -e` 환경에서 스크립트가 중단될 수 있었습니다. `|| true`로 올바르게 수정되었습니다.

File: scripts/setup-cron.sh [hunk 7]
```diff
@@ -1,26 +1,41 @@
...
-echo "0 9 * * * cd $PROJECT_DIR && export PATH=/usr/local/bin:/usr/bin:\$PATH && [ -f .env ] && set -a && . ./.env && set +a && ./scripts/log-watcher.sh >> $LOG_DIR/ai-monitor.log 2>&1" >> /tmp/crontab.tmp
+# 서버 로그 디렉토리 (logging.rs의 LOG_DIR과 일치시킴)
+SERVER_LOG_DIR="$PROJECT_DIR/codes/server/logs"
+mkdir -p "$SERVER_LOG_DIR"
+
+# 환경변수 로드 + PATH 설정 공통 prefix
+ENV_PREFIX="cd $PROJECT_DIR && export PATH=/usr/local/bin:/usr/bin:\$PATH && export LOG_DIR=$SERVER_LOG_DIR && [ -f .env ] && set -a && . ./.env && set +a"
+
+# 1. 에러 AI 진단 (매일 9시)
+echo "0 9 * * * $ENV_PREFIX && ./scripts/log-watcher.sh >> $LOG_DIR/ai-monitor.log 2>&1" >> /tmp/crontab.tmp
```
Comment (jsoonworld) [inline scripts/setup-cron.sh:22]: daily-report.sh를 9:05에 실행하여 log-watcher.sh(9:00)와의 경쟁 조건을 피한 것은 좋은 설계입니다.

**Review by jsoonworld — COMMENTED** [review]
모니터링 기능 활성화 PR 리뷰입니다. 전반적으로 잘 구성되어 있습니다. 몇 가지 개선 제안을 드립니다.

File: codes/server/src/utils/logging.rs (hunk 2 above)
Comment (catturtle123) [reply codes/server/src/utils/logging.rs:27]: 수정 완료했습니다. (커밋 357166a)

`init_logging()`이 `WorkerGuard`를 반환하도록 변경하고, `main()`에서 `_log_guard`로 유지합니다:

```rust
// logging.rs
pub fn init_logging() -> tracing_appender::non_blocking::WorkerGuard {
    let (non_blocking, guard) = tracing_appender::non_blocking(file_appender);
    // ...
    guard
}

// main.rs
let _log_guard = utils::init_logging();
```

`non_blocking` writer를 사용하므로 파일 쓰기가 별도 스레드에서 처리되어 성능도 개선됩니다.

File: scripts/daily-report.sh (hunk 3 above)
Comment (catturtle123) [reply scripts/daily-report.sh:79]: 수정 완료했습니다. (커밋 357166a)

섹션 제목을 "**현재 시스템 상태**"로 변경하여 어제 로그 분석 데이터가 아닌 현재 스냅샷임을 명시했습니다.

File: scripts/perf-alert.sh (hunk 4 above)
Comment (catturtle123) [reply scripts/perf-alert.sh:90]: 수정 완료했습니다. (커밋 357166a)

`bc` 의존성을 제거하고 bash 정수 비교로 대체했습니다:

```bash
CPU_INT=${CPU_USAGE%.*}  # 소수점 제거하여 정수 비교
if [ "${CPU_INT:-0}" -ge "$CPU_THRESHOLD" ]; then
```

메모리 비교도 동일하게 변경했습니다.

File: scripts/perf-alert.sh (hunk 5 above)
Comment (catturtle123) [reply scripts/perf-alert.sh:138]: 리뷰 감사합니다. 말씀하신 대로 현재 분 단위 매칭은 대략적인 모니터링 용도로 충분하다고 판단하여 현재 구현을 유지합니다. 추후 정밀한 에러 추적이 필요해지면 `jq`로 timestamp 파싱하는 방식으로 개선하겠습니다.

File: scripts/run-tests.sh (hunk 6 above)
Comment (catturtle123) [reply scripts/run-tests.sh:16]: 감사합니다! 👍

File: scripts/setup-cron.sh (hunk 7 above)
Comment (catturtle123) [reply scripts/setup-cron.sh:22]: 감사합니다!

File: codes/server/src/utils/logging.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (catturtle123) [reply codes/server/src/utils/logging.rs:21]: 추가 수정 완료했습니다. (커밋 357166a)

1. **파일명 형식**: 이전 커밋에서 doc comment를 `server.log.YYYY-MM-DD`로 수정 완료
2. **셸 스크립트**: `daily-report.sh`, `perf-alert.sh`, `log-watcher.sh` 모두 `server.log.$DATE` 형식으로 통일 확인됨
3. **WorkerGuard**: `init_logging()`이 `WorkerGuard`를 반환하도록 변경하고, `main()`에서 `_log_guard`로 유지하여 종료 시 버퍼링된 로그 flush 보장

또한 `non_blocking` writer를 사용하여 파일 I/O가 별도 스레드에서 처리됩니다.

File: codes/server/src/utils/logging.rs [hunk 8]
```diff
@@ -39,4 +59,6 @@ pub fn init_logging() {
             Err(err)
         })
         .ok(); // Convert to unit, letting the server start even if logging fails
+
+    guard
```
Comment (jsoonworld) [inline codes/server/src/utils/logging.rs:63]: [Bug] `init_logging()`에서 `try_init()`이 실패할 경우(이미 초기화된 경우 등), `.ok()`로 에러를 무시하고 `guard`를 반환합니다. 그런데 이 경우 `file_layer`가 실제로 등록되지 않았음에도 `guard`가 반환되어, `main()`에서는 로깅이 정상 작동하는 것처럼 보이지만 실제로는 파일 로깅이 동작하지 않는 상황이 발생할 수 있습니다.

현재 프로덕션에서는 `init_logging()`이 한 번만 호출되므로 큰 문제는 아니지만, 테스트 환경에서 여러 번 호출될 때 혼란을 줄 수 있습니다. 향후 개선 시 참고해주세요.

File: scripts/perf-alert.sh [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
```
Comment (jsoonworld) [inline scripts/perf-alert.sh:138]: [Suggestion] `perf-alert.sh`에서 에러 급증을 감지할 때 `grep "\"level\":\"ERROR\""`로 로그 레벨을 확인하고 있습니다. 그런데 미들웨어의 `error!()` 매크로는 `tracing-subscriber` JSON 포맷에서 `"level":"ERROR"`로 출력됩니다. 이 부분은 정상입니다.

다만, 미들웨어에서 5xx 응답을 `error!()`로 기록하면 `perf-alert.sh`에서는 이를 에러 급증으로 감지하지만, `log-watcher.sh`는 `error_code` 필드를 기반으로 심각도를 판단합니다. 미들웨어의 `error!()` 로그에는 `error_code` 필드가 없으므로 `log-watcher.sh`에서는 `UNKNOWN` 코드로 분류되어 `info` 심각도로 처리됩니다.

즉, 같은 5xx 에러가 `perf-alert.sh`에서는 감지되지만 `log-watcher.sh`에서는 무시될 수 있습니다. 의도된 동작이라면 괜찮지만, 명확히 해두면 좋겠습니다.

File: scripts/perf-alert.sh [hunk 10]
```diff
@@ -0,0 +1,134 @@
+#!/bin/bash
+# scripts/perf-alert.sh - 즉시 성능 알림
+# 매분 Cron으로 실행되어 임계값 초과 시 Discord 알림 전송
+# 경량 스크립트: AI 호출 없음, 시스템 지표만 체크
+
+set -euo pipefail
+
```
Comment (jsoonworld) [inline scripts/perf-alert.sh:7]: [Suggestion] `check_dedup()` 함수에서 dedup 파일을 읽고 업데이트할 때, grep + mv 패턴을 사용하고 있는데, 동시에 여러 metric 체크가 같은 파일에 접근하면 race condition이 발생할 수 있습니다. 

현재 구현에서는 스크립트가 순차적으로 CPU -> MEM -> ERROR 순서로 실행되므로 실제로는 문제가 없지만, cron이 매분 실행되면서 이전 실행이 아직 끝나지 않은 경우 두 인스턴스가 동시에 실행될 수 있습니다.

간단한 해결: 스크립트 시작 부분에 `flock`이나 PID file 기반의 동시 실행 방지 로직을 추가하면 안전합니다.

```bash
LOCK_FILE="/tmp/perf-alert.lock"
exec 200>"$LOCK_FILE"
flock -n 200 || { echo "Already running"; exit 0; }
```

다만 이 스크립트는 매우 가볍고 빠르게 종료되므로, 현실적으로 겹칠 확률은 낮습니다.

File: scripts/daily-report.sh [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    echo "[$(date)] ERROR: jq is required but not installed" >&2
+    exit 1
+fi
+
+# duration_ms 추출 (request completed 로그에서)
+DURATIONS=$(jq -r 'select(.fields.message == "request completed") | .fields.duration_ms // empty' "$LOG_FILE" 2>/dev/null | sort -n)
+
+if [ -z "$DURATIONS" ]; then
+    echo "[$(date)] INFO: No request logs found for $YESTERDAY"
+    exit 0
```
Comment (jsoonworld) [inline scripts/daily-report.sh:37]: [Bug] daily-report.sh에서 jq로 로그를 파싱할 때, 전체 로그 파일의 duration_ms 값을 셸 변수 DURATIONS에 로드합니다. 프로덕션에서 하루 요청이 많아질 경우(예: 수만 건), 메모리 사용량이 증가할 수 있습니다. 현재 트래픽 수준에서는 문제가 없겠지만, 향후 스케일링 시에는 임시 파일을 사용하는 방식으로 개선을 고려해보세요.

File: scripts/setup-cron.sh [hunk 12]
```diff
@@ -1,26 +1,41 @@
 #!/bin/bash
 # scripts/setup-cron.sh - Cron job 설정
+# 모니터링 스크립트들의 스케줄을 등록합니다.
 
 PROJECT_DIR="$(cd "$(dirname "$0")/.." && pwd)"
 LOG_DIR="$PROJECT_DIR/logs"
```
Comment (jsoonworld) [inline scripts/setup-cron.sh:6]: [Question] setup-cron.sh에서 cron 실행 로그(ai-monitor.log, daily-report.log 등)는 $PROJECT_DIR/logs/ 에 저장되고, 서버 로그(server.log.YYYY-MM-DD)는 $PROJECT_DIR/codes/server/logs/ 에 저장됩니다. 로그 디렉토리가 두 곳으로 분리된 것이 의도된 설계인지 궁금합니다. 서버 로그와 모니터링 로그의 성격이 다르니 분리하는 것이 합리적이지만, 운영 시 로그 관리 포인트가 늘어날 수 있습니다. 로그 보존/정리 정책(예: logrotate)도 추후 고려해보시면 좋겠습니다.

File: codes/server/src/global/middleware.rs [hunk 13]
```diff
@@ -19,19 +18,44 @@ pub async fn request_id_middleware(mut request: Request, next: Next) -> Response
         .extensions_mut()
         .insert(RequestId(request_id.clone()));
 
-    // Span에 request_id 포함 - instrument()로 async-safe하게 적용
+    let method = request.method().to_string();
```
Comment (jsoonworld) [inline codes/server/src/global/middleware.rs]: [Nit] 미들웨어에서 method와 path를 String으로 변환하여 캡처하고 있는데, span에서 request가 move되기 전에 값을 복사하기 위한 것으로 이해됩니다. 잘 처리되어 있습니다. 한 가지 작은 개선: method는 HTTP Method enum이므로 to_string() 대신 as_str()을 사용하면 불필요한 allocation을 줄일 수 있습니다. 다만 span의 lifetime 때문에 현재처럼 owned String이 필요한 상황이어서, 현재 구현이 충분히 올바릅니다.

**Review by jsoonworld — COMMENTED** [review]
## 코드 리뷰 요약

전반적으로 모니터링 파이프라인이 잘 구성되어 있습니다. jsoonworld님이 이전에 지적하신 사항들(WorkerGuard, bc 의존성, 리포트 제목)이 모두 잘 반영되었습니다.

### 잘된 점
- **`tracing-appender`의 `non_blocking` writer 사용**: 파일 I/O가 별도 스레드에서 처리되어 요청 처리 성능에 영향을 주지 않습니다
- **`WorkerGuard` 반환 패턴**: main에서 guard를 유지하여 프로세스 종료 시 버퍼링된 로그가 안전하게 flush됩니다
- **미들웨어의 로그 레벨 분리 (info/warn/error)**: status code 기반으로 적절히 분류되어 있습니다
- **셸 스크립트의 크로스 플랫폼 지원** (Darwin/Linux): EC2(Linux)와 로컬(macOS) 모두에서 동작합니다
- **`set -euo pipefail`과 `|| true` 패턴**: 파이프라인 에러 처리가 견고합니다
- **Cron 스케줄 설계**: daily-report를 log-watcher 5분 후에 실행하여 경쟁 조건을 방지한 것이 좋습니다
- **`run-tests.sh`의 `|| true` 수정**: `grep -c`가 0일 때 exit code 1을 반환하는 문제를 올바르게 수정했습니다

### 개선 제안 사항 (라인 코멘트 참조)
1. **[logging.rs]** `try_init()` 실패 시에도 guard가 반환되는 edge case
2. **[perf-alert.sh]** 에러 감지 로직이 log-watcher.sh와 다른 기준으로 동작하는 점 (error_code 유무)
3. **[perf-alert.sh]** 동시 실행 방지를 위한 flock 사용 제안
4. **[daily-report.sh]** 대용량 로그 시 셸 변수 메모리 사용 우려
5. **[setup-cron.sh]** 로그 디렉토리 이원화에 대한 질문

### 블로킹 이슈 없음
위 사항들은 모두 향후 개선을 위한 제안이며, 현재 머지에 블로킹되는 이슈는 없습니다. 좋은 PR 감사합니다!

**Review by jsoonworld — COMMENTED** [review]
## Code Review - 모니터링 활성화 및 P95/P99 리포트

### 수정 완료
- ✅ `.gitignore`에 `logs/` 추가 (로그 파일 커밋 방지)
- ✅ `.env.example`에 `LOG_DIR` 환경변수 추가

### Rust 코드 품질
- ✅ unwrap() 사용 없음, tracing 적절 사용
- ✅ WorkerGuard 패턴 정상 구현
- ✅ 비동기 파일 I/O에 non_blocking writer 사용

### 개선 제안 (후속 작업 가능)
- 💡 `perf-alert.sh` 상태 저장을 `/tmp` 대신 프로젝트 상대경로로 변경 권장
- 💡 에러 burst 감지 타임스탬프 매칭 개선 권장
- 💡 임계값을 환경변수로 설정 가능하게 변경 권장

File: codes/server/src/utils/logging.rs (hunk 8 above)
Comment (catturtle123) [reply codes/server/src/utils/logging.rs:63]: 좋은 지적입니다! 현재 코드에서는 `try_init()` 실패 시에도 `guard`를 반환하고 있는데, 이는 의도적인 설계입니다.

`guard`는 `tracing_appender::non_blocking()`에서 생성되며, subscriber 초기화와는 독립적입니다. `guard`를 유지하지 않으면 non-blocking writer의 백그라운드 스레드가 종료되어 버퍼링된 로그가 손실될 수 있습니다.

다만, `try_init()` 실패 시 `eprintln!`으로 경고를 출력하고 있어 운영자가 문제를 인지할 수 있습니다. 추가로 `tracing::dispatcher::SetGlobalDefaultError`(이미 초기화된 경우)는 무시하도록 처리되어 있습니다.

File: scripts/perf-alert.sh (hunk 9 above)
Comment (catturtle123) [reply scripts/perf-alert.sh:138]: 좋은 지적입니다! 현재 middleware의 `error!()` 로그에는 `error_code` 필드가 없어서 log-watcher가 이를 `UNKNOWN`으로 분류하고, `get_error_severity("UNKNOWN")`이 `info`를 반환하여 알림이 발송되지 않는 문제가 있습니다.

이를 개선하려면 두 가지 접근이 가능합니다:

1. **middleware에서 status 기반 error_code 추가**: 5xx → `COMMON500` 등으로 매핑
2. **log-watcher에서 UNKNOWN + status >= 500 조합 처리**: `error_code`가 없지만 `status` 필드가 500 이상이면 critical로 분류

현재 perf-alert.sh의 에러 급증 체크가 ERROR 레벨 로그 전체를 카운트하므로 1차 안전망 역할을 하고 있지만, log-watcher와의 일관성을 위해 후속 이슈로 개선하겠습니다.

File: scripts/perf-alert.sh (hunk 10 above)
Comment (catturtle123) [reply scripts/perf-alert.sh:7]: 반영 완료했습니다! flock 기반 동시 실행 방지를 추가했습니다.

- Linux: `flock -n` (non-blocking)으로 락 획득 시도, 실패 시 즉시 종료
- macOS: `mkdir` atomic operation으로 fallback (flock이 없는 환경 대응)
- 락 획득 실패 시 로그 출력 후 `exit 0`으로 정상 종료 (Cron 에러 메일 방지)

log-watcher.sh에서 이미 사용하고 있던 패턴과 동일한 방식으로 구현했습니다.

File: scripts/daily-report.sh (hunk 11 above)
Comment (catturtle123) [reply scripts/daily-report.sh:37]: 맞습니다! 현재 `DURATIONS` 변수가 전체 duration 값을 메모리에 로드하는 구조입니다.

운영 환경에서 일일 요청량이 매우 많아질 경우 (수십만 건 이상) 메모리 이슈가 될 수 있습니다. 개선 방안으로는:

1. **awk 스트리밍 처리**: `jq` + `awk`로 한 번에 총 요청 수, P50/P95/P99를 계산 (메모리 O(1))
2. **임시 파일 활용**: duration 값을 임시 파일에 저장 후 처리

현재 서비스 규모에서는 문제없지만, 트래픽 증가 시 개선이 필요한 부분으로 인지하고 있겠습니다.

File: scripts/setup-cron.sh (hunk 12 above)
Comment (catturtle123) [reply scripts/setup-cron.sh:6]: 좋은 질문입니다! 현재 로그 디렉토리는 이미 분리되어 있습니다:

- **서버 로그**: `$PROJECT_DIR/codes/server/logs/` (logging.rs의 LOG_DIR, server.log.YYYY-MM-DD 형태)
- **모니터링 로그**: `$PROJECT_DIR/logs/` (setup-cron.sh에서 각 스크립트의 stdout/stderr 리다이렉션)

setup-cron.sh의 6번 라인 `LOG_DIR`은 모니터링 스크립트 자체의 실행 로그를 저장하는 디렉토리이고, 15번 라인 `SERVER_LOG_DIR`은 서버가 생성하는 애플리케이션 로그를 가리킵니다. 각 모니터링 스크립트에는 `LOG_DIR` 환경변수로 서버 로그 경로를 전달합니다.

File: codes/server/src/global/middleware.rs (hunk 13 above)
Comment (catturtle123) [reply codes/server/src/global/middleware.rs]: 반영했습니다! `request.method().to_string()`을 `request.method().as_str()`로 변경했습니다.

`Method::as_str()`은 내부 문자열을 직접 반환하므로 `to_string()`의 Display → format! → String 할당 과정을 건너뛸 수 있습니다.

span의 `method = %method` 구문에서 `%`는 생성 시점에 Display 포맷을 기록하므로, `&str` 값이 span에 복사된 후에는 원본 참조가 필요 없습니다. NLL 덕분에 이후 `request`를 async 블록으로 이동해도 borrow 충돌이 발생하지 않아 컴파일이 정상적으로 통과합니다.

**Comment by catturtle123** [comment]
@coderabbitai review

**Review by jsoonworld — APPROVED** [review]
